packages/vats/src/nameHub.js

```diff
@@ -1,215 +1,360 @@
-      return harden([
-        ...mapIterable(
-          keyToRecord.entries(),
-          ([key, record]) =>
-            /** @type {[string, ERef<unknown>]} */ ([
-              key,
-              record.promise || record.value,
-            ]),
-        ),
-      ]);
-    },
-    values: () => {
-      return [
-        ...mapIterable(
-          keyToRecord.values(),
-          record => record.promise || record.value,
-        ),
-      ];
-    },
-    keys: () => {
-      return harden([...keyToRecord.keys()]);
-    },
+export const prepareNameHubKit = (zone = heapZone) => {
+  const init1 = () => ({
+    /** @type {LegacyMap<string, PromiseKit<unknown>>} */
+    // Legacy because a promiseKit is not a passable
+    keyToPK: makeLegacyMap('nameKey'),
+    /** @type {LegacyMap<string, PromiseKit<unknown>>} */
+    // Legacy because a promiseKit is not a passable
+    keyToAdminPK: makeLegacyMap('nameKey'),
   });
+  /** @type {WeakMap<any, ReturnType<init1>>} */
+  const ephemera = new WeakMap();
+  /** @param {{}} me */
+  const my = me => provideWeak(ephemera, me, init1);
 
-  /** @type {LegacyMap<string, NameRecord>} */
-  // Legacy because a promiseKit is not a passable
-  const keyToAdminRecord = makeLegacyMap('nameKey');
-  /** @type {undefined | ((entries: [string, unknown][]) => ERef<void>)} */
-  let updateCallback;
-  const updated = () => {
-    let result;
-    if (updateCallback) {
-      result = updateCallback(
-        harden(
-          [
-            ...mapIterable(keyToRecord.entries(), ([name, record]) =>
-              record.promise
-                ? []
-                : [/** @type {[string, unknown]} */ ([name, record.value])],
+  /** @type {() => import('./types').NameHubKit } */
+  const makeNameHubKit = zone.exoClassKit(
+    'NameHubKit',
+    NameHubIKit,
+    /** @param {unknown[]} auxProperties */
+    (...auxProperties) => ({
+      /** @type {MapStore<string, unknown>} */
+      keyToValue: zone.detached().mapStore('nameKey'),
+
+      /** @type {MapStore<string, import('./types').NameAdmin>} */
+      keyToAdmin: zone.detached().mapStore('nameKey'),
+
+      /** @type {undefined | { write: (item: unknown) => void }} */
+      updateCallback: undefined,
+
+      auxProperties: harden(auxProperties),
+    }),
+    {
+      nameHub: {
+        has(key) {
+          const { keyToValue } = this.state;
+          const { keyToPK } = my(this.facets.nameHub);
+          return keyToValue.has(key) || keyToPK.has(key);
+        },
+        async lookup(...path) {
+          if (path.length === 0) {
+            const { nameHub } = this.facets;
+            return nameHub;
+          }
+          const { keyToValue } = this.state;
+          const { keyToPK } = my(this.facets.nameHub);
+          const [first, ...remaining] = path;
+          /** @type {any} */
+          const firstValue = keyToValue.has(first)
+            ? keyToValue.get(first)
+            : keyToPK.get(first).promise; // or throw
```

Comment:
JS Map doesn't throw on missing values. To get that back a clean option is our `NonNullish` checked cast:
```suggestion
            : NonNullish(keyToPK.get(first)).promise;
```